[FLINK-13127] Fix --yarnship classpath

dmvk opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Travis CI to do that following [this guide](http://flink.apache.org/contribute-code.html#best-practices).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

Fix class loading of resources shipped via `--yarnship <files>`. More details in [jira issue](https://issues.apache.org/jira/browse/FLINK-13127).

## Brief change log

  - *AbstractYarnClusterDescriptor* adds directories to classpath instead of files in case it finds non-jar files

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

Tested on private yarn cluster.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

TisonKun commented on flink-yarn/src/main/java/org/apache/flink/yarn/AbstractYarnClusterDescriptor.java:
```diff
@@ -1162,7 +1172,11 @@ public FileVisitResult visitFile(java.nio.file.Path file, BasicFileAttributes at
 
 								// add files to the classpath
-								classPaths.add(key);
+								if (fileName.endsWith("jar")) {
```
point 2 refers these following 5 lines.

TisonKun commented on flink-yarn/src/main/java/org/apache/flink/yarn/AbstractYarnClusterDescriptor.java:
```diff
@@ -1181,11 +1195,21 @@ public FileVisitResult visitFile(java.nio.file.Path file, BasicFileAttributes at
 
 					// add files to the classpath
-					classPaths.add(key);
+					if (key.endsWith("jar")) {
```
and these following 5 lines.

It could be `addFilesToClassPath(classPath, shipFile)`

xintongsong commented:
Hi @dmvk,
Thanks for opening this PR. I think this is a helpful contribution. A few comments:
1. I think it would be better if you could find a committer assigning this jira ticket to you before starting the implementation, so other contributors know you are working on this and will not start a redundant work. @tillrohrmann ,could you help assigning the jira ticket?
2. Agree with @TisonKun, I think this change is not that trivial. It would be better to have a test case to make sure other people won't accidentally break it in the future.
3. Why do we need to separate and sort archives and resources? I mean, I'm total fine with this because it makes the classpath organized and friendly for people to look into. Just curious whether there is any other reason that makes this necessary.

dmvk commented:
@xintongsong @TisonKun thanks for the review! For points 1) and 2), I've updated a test case and removed duplicated code in `org.apache.flink.yarn.AbstractYarnClusterDescriptor#uploadAndRegisterFiles` method.

I don't think there is any significant benefit of sorting those, the idea was to make sure that it is deterministic, but that would be probably the case without sorting anyway. I don't have any strong arguments for this, so I can remove it if you want to.

Thanks!

TisonKun commented on flink-yarn/src/test/java/org/apache/flink/yarn/YarnFileStageTest.java:
```diff
@@ -177,14 +179,23 @@ static void testCopyFromLocalRecursive(
+					srcDir.getName() + "/nested",
+					srcDir.getName() + "/nested/4",
+					srcDir.getName() + "/test.jar")
+				, classpath);
```
FYI, you can make use of `Assert.assertThat` and `Matchers.containsInAnyOrder`.

Here is an example 

```java
Assert.assertThat(new ArrayList<Integer>() {{
			add(1); add(2); add(3);
		}}, Matchers.containsInAnyOrder(3, 2, 1));
```

xintongsong commented:
Maybe also rebase to the latest master? There is a travis test failure, which is not caused by this PR and should been fixed on master branch.

dmvk commented:
@xintongsong done ;) thanks

xintongsong commented:
@dmvk Thanks for addressing the comments, the PR looks good to me.
@tillrohrmann Could you take a final check on this?

dmvk commented:
@tillrohrmann did you have a time to take a look at this?

tillrohrmann commented:
I'm looking at it right now.

dmvk commented:
@tillrohrmann any progress? thanks

tillrohrmann commented:
The test failure is unrelated. Thanks a lot for the review @TisonKun and @xintongsong.
